src/lib/operators/join_nested_loop.cpp

```diff
@@ -91,39 +92,38 @@ void JoinNestedLoop::_join_two_untyped_segments(const std::shared_ptr<const Base
                                                 const std::shared_ptr<const BaseSegment>& segment_right,
                                                 const ChunkID chunk_id_left, const ChunkID chunk_id_right,
                                                 JoinNestedLoop::JoinParams& params) {
-  resolve_data_and_segment_type(*segment_left, [&](auto left_type, auto& typed_left_segment) {
-    resolve_data_and_segment_type(*segment_right, [&](auto right_type, auto& typed_right_segment) {
-      using LeftType = typename decltype(left_type)::type;
-      using RightType = typename decltype(right_type)::type;
-
-      // make sure that we do not compile invalid versions of these lambdas
-      constexpr auto LEFT_IS_STRING_COLUMN = (std::is_same<LeftType, std::string>{});
-      constexpr auto RIGHT_IS_STRING_COLUMN = (std::is_same<RightType, std::string>{});
-
-      constexpr auto NEITHER_IS_STRING_COLUMN = !LEFT_IS_STRING_COLUMN && !RIGHT_IS_STRING_COLUMN;
-      constexpr auto BOTH_ARE_STRING_COLUMN = LEFT_IS_STRING_COLUMN && RIGHT_IS_STRING_COLUMN;
-
-      if constexpr (NEITHER_IS_STRING_COLUMN || BOTH_ARE_STRING_COLUMN) {
-        auto iterable_left = create_iterable_from_segment<LeftType>(typed_left_segment);
-        auto iterable_right = create_iterable_from_segment<RightType>(typed_right_segment);
-
-        // Dirty hack to avoid https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86740
-        const auto params_copy = params;
-        const auto chunk_id_left_copy = chunk_id_left;
-        const auto chunk_id_right_copy = chunk_id_right;
-
-        iterable_left.with_iterators(
-            [&params_copy, &iterable_right, chunk_id_left_copy, chunk_id_right_copy](auto left_it, auto left_end) {
-              iterable_right.with_iterators([&](auto right_it, auto right_end) {
-                with_comparator(params_copy.predicate_condition, [&](auto comparator) {
-                  join_two_typed_segments(comparator, left_it, left_end, right_it, right_end, chunk_id_left_copy,
-                                          chunk_id_right_copy, params_copy);
-                });
-              });
+  segment_with_iterators<ResolveDataTypeTag, SegmentIterationTypeErasure::Always>(
```

Comment:
This code is also used as a fallback by JoinIndex. As that operator is not used by default right now, this is not an issue, but we should probably add a comment over there saying that even for the nested loop join, this is not the most efficient way to do things.

Reply:
```
  /**
   * The nested loops.
   *
   * The value in the outer loop is retrieved via virtual function calls ("SegmentIterationTypeErasure::Always") and only
   * the inner loop gets inlined. This is to keep the compile time of the JoinNestedLoop *somewhat* at bay, if we inline
   * both the inner and the outer loop, the JoinNestedLoop becomes the most expensive-to-compile file in all of Hyrise
   * by a margin
   */
```

Glad to discuss the JNL in person. As it stands it's a major pain compile time wise and - as you say - we don't even use it right now. We could consider other solutions, including materializing both columns before joining. This would actually avoid repeatedly de-coding the inner relation....

Reply:
For the index join, we should find a better solution anyway. Using the nested loop join here is just a quick fix.